File: api/extract.js
/**
 * api/extract.js — Vercel Node.js Serverless Function
 * POST /api/extract  { url, audio_only }
 *
 * Menggantikan versi Python (yt-dlp) sebelumnya. Sekarang setiap platform
 * punya beberapa metode ekstraksi berurutan (kalau metode pertama gagal,
 * otomatis coba metode berikutnya) — lihat api/_lib/platforms/*.js.
 * Kontrak request/response TIDAK berubah, jadi assets/main.js dan semua
 * halaman *-downloader.html tetap jalan tanpa modifikasi.
 */

const { extractTiktok } = require('./_lib/platforms/tiktok');
const { extractInstagram } = require('./_lib/platforms/instagram');
const { extractFacebook } = require('./_lib/platforms/facebook');
const { extractTwitter } = require('./_lib/platforms/twitter');
const { extractYoutube } = require('./_lib/platforms/youtube');
const { toProxyUrl } = require('./_lib/proxy-url');

const PLATFORM_PATTERNS = {
  tiktok: /tiktok\.com/i,
  instagram: /instagram\.com/i,
  facebook: /facebook\.com|fb\.watch/i,
  x: /(?:twitter|x)\.com/i,
  youtube: /youtube\.com|youtu\.be/i,
};

const EXTRACTORS = {
  tiktok: extractTiktok,
  instagram: extractInstagram,
  facebook: extractFacebook,
  x: extractTwitter,
  youtube: extractYoutube,
};

function detectPlatform(url) {
  for (const [name, pattern] of Object.entries(PLATFORM_PATTERNS)) {
    if (pattern.test(url)) return name;
  }
  return null;
}

function setCors(res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
}

function send(res, status, body) {
  res.status(status).setHeader('Content-Type', 'application/json');
  res.end(JSON.stringify(body));
}

function friendlyError(platform, err) {
  const msg = (err && err.message) || String(err);
  if (platform === 'instagram' && /private|login/i.test(msg)) {
    return 'Konten ini kemungkinan private atau butuh login. Pastikan post/reel-nya publik.';
  }
  return `Gagal memproses link. Pastikan kontennya publik dan link valid. (${msg})`;
}

module.exports = async (req, res) => {
  setCors(res);

  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }
  if (req.method === 'GET') {
    send(res, 200, { status: 'ok' });
    return;
  }
  if (req.method !== 'POST') {
    send(res, 405, { detail: 'Method tidak didukung.' });
    return;
  }

  let payload = {};
  try {
    payload = typeof req.body === 'object' && req.body ? req.body : JSON.parse(req.body || '{}');
  } catch {
    payload = {};
  }

  const url = (payload.url || '').trim();
  const audioOnly = Boolean(payload.audio_only);

  if (!url) {
    send(res, 400, { detail: "Field 'url' wajib diisi." });
    return;
  }

  const platform = detectPlatform(url);
  if (!platform) {
    send(res, 400, { detail: 'Link tidak dikenali. Dukung: TikTok, Instagram, Facebook, X, YouTube.' });
    return;
  }

  try {
    const result = await EXTRACTORS[platform](url, { audioOnly });

    let formats = result.formats || [];
    if (audioOnly) {
      const audioOnly_ = formats.filter((f) => /audio|mp3/i.test(f.label));
      if (audioOnly_.length) formats = audioOnly_;
    }
    if (!formats.length) {
      send(res, 422, { detail: 'Tidak ditemukan format yang bisa diunduh untuk link ini.' });
      return;
    }

    // Bungkus tiap link lewat /api/media supaya referer/hotlink-block
    // teratasi dan file benar-benar ke-download (bukan cuma kebuka di tab).
    const wrappedFormats = formats.slice(0, 6).map((f, i) => ({
      format_id: String(i),
      ext: audioOnly ? 'mp3' : 'mp4',
      label: f.label,
      filesize_approx: f.filesize_approx || null,
      url: toProxyUrl(f.url, platform, `toolkitme-${platform}-${i + 1}`),
    }));

    send(res, 200, {
      platform,
      title: result.title || 'Tanpa judul',
      thumbnail: result.thumbnail || null,
      duration: result.duration || null,
      uploader: result.uploader || null,
      formats: wrappedFormats,
    });
  } catch (err) {
    send(res, 422, { detail: friendlyError(platform, err) });
  }
};

File: api/extract.py
"""
api/extract.py — Vercel Python Serverless Function

Vercel otomatis mengubah file ini jadi endpoint: POST /api/extract
Tidak perlu server terpisah, tidak perlu Railway — jalan di project
Vercel yang sama dengan index.html.
"""

import json
import re
from http.server import BaseHTTPRequestHandler

import yt_dlp

PLATFORM_PATTERNS = {
    "tiktok": r"tiktok\.com",
    "instagram": r"instagram\.com",
    "facebook": r"facebook\.com|fb\.watch",
    "x": r"(?:twitter|x)\.com",
    "youtube": r"youtube\.com|youtu\.be",
}


def detect_platform(url: str):
    for name, pattern in PLATFORM_PATTERNS.items():
        if re.search(pattern, url, re.IGNORECASE):
            return name
    return None


def build_ydl_opts(audio_only: bool) -> dict:
    return {
        "quiet": True,
        "no_warnings": True,
        "skip_download": True,
        "noplaylist": True,
        "format": "bestaudio/best" if audio_only else "best",
    }


def run_extract(url: str, audio_only: bool):
    platform = detect_platform(url)
    if not platform:
        return 400, {"detail": "Link tidak dikenali. Dukung: TikTok, Instagram, Facebook, X, YouTube."}

    try:
        with yt_dlp.YoutubeDL(build_ydl_opts(audio_only)) as ydl:
            info = ydl.extract_info(url, download=False)
    except yt_dlp.utils.DownloadError as e:
        return 422, {"detail": f"Gagal memproses link. Pastikan kontennya publik dan link valid. ({e})"}
    except Exception as e:
        return 500, {"detail": f"Kesalahan server: {e}"}

    if info is None:
        return 422, {"detail": "Tidak ada data yang bisa diambil dari link ini."}

    if "entries" in info:
        info = info["entries"][0]

    raw_formats = info.get("formats") or [info]
    formats = []
    for f in raw_formats:
        direct_url = f.get("url")
        if not direct_url:
            continue
        if audio_only and f.get("vcodec") not in (None, "none"):
            continue
        if not audio_only and f.get("vcodec") in (None, "none"):
            continue
        height = f.get("height")
        formats.append({
            "format_id": str(f.get("format_id", "")),
            "ext": f.get("ext", "mp4" if not audio_only else "mp3"),
            "label": f"{height}p" if height else f.get("format_note", f.get("ext", "file")),
            "filesize_approx": f.get("filesize") or f.get("filesize_approx"),
            "url": direct_url,
        })

    if not formats and info.get("url"):
        formats.append({
            "format_id": "direct",
            "ext": info.get("ext", "mp4"),
            "label": "Original",
            "filesize_approx": None,
            "url": info["url"],
        })

    if not formats:
        return 422, {"detail": "Tidak ditemukan format yang bisa diunduh untuk link ini."}

    return 200, {
        "platform": platform,
        "title": info.get("title") or "Tanpa judul",
        "thumbnail": info.get("thumbnail"),
        "duration": info.get("duration"),
        "uploader": info.get("uploader") or info.get("channel"),
        "formats": formats,
    }


class handler(BaseHTTPRequestHandler):
    def _cors(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")

    def do_OPTIONS(self):
        self.send_response(200)
        self._cors()
        self.end_headers()

    def do_GET(self):
        self.send_response(200)
        self._cors()
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps({"status": "ok"}).encode())

    def do_POST(self):
        length = int(self.headers.get("Content-Length", 0))
        raw = self.rfile.read(length) if length else b"{}"
        try:
            payload = json.loads(raw or b"{}")
        except json.JSONDecodeError:
            payload = {}

        url = (payload.get("url") or "").strip()
        audio_only = bool(payload.get("audio_only", False))

        if not url:
            status, body = 400, {"detail": "Field 'url' wajib diisi."}
        else:
            status, body = run_extract(url, audio_only)

        self.send_response(status)
        self._cors()
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps(body).encode())
